toml/src/main/kotlin/org/rust/toml/crates/local/CratesLocalIndexService.kt

```diff
@@ -0,0 +1,236 @@
+
+    @Volatile
+    var isReady: Boolean = true
+
+    override fun getState(): CratesLocalIndexState = _state.get()
+    override fun loadState(state: CratesLocalIndexState) {
+        _state.set(state)
+    }
+
+    private fun updateIfNeeded() {
+        if (state.indexedCommitHash != registryHeadCommitHash && isReady) {
+            isReady = false
+            CratesLocalIndexUpdateTask().queue()
+        }
+    }
+
+    private fun resetIndexIfNeeded() {
+        if (corruptionMarkerFile.exists()) resetIndex()
+    }
+
+    private fun resetIndex() {
+        baseCratesLocalRegistryDir.cleanDirectory()
+        state.indexedCommitHash = ""
+    }
+
+    fun getCrate(crateName: String): CargoRegistryCrate? {
+        updateIfNeeded()
+
+        return crates.get(crateName)
+    }
+
+    fun getAllCrateNames(): List<String> {
+        updateIfNeeded()
+
+        val crateNames = mutableListOf<String>()
+        crates.processKeys { name ->
+            crateNames.add(name)
+        }
+
+        return crateNames
+    }
+
+    fun reloadCrates(revTree: RevTree) {
+        TreeWalk(repository).use { treeWalk ->
+            treeWalk.addTree(revTree)
+            treeWalk.isRecursive = true
+            treeWalk.isPostOrderTraversal = false
+
+            while (treeWalk.next()) {
+                if (treeWalk.pathString == "config.json") continue
+
+                val objectId = treeWalk.getObjectId(0)
+                val loader = repository.open(objectId)
+                val versions = mutableListOf<CargoRegistryCrateVersion>()
+                val reader = BufferedReader(InputStreamReader(loader.openStream()))
+
+                reader.forEachLine { line ->
+                    if (line.isBlank()) return@forEachLine
+
+                    try {
+                        versions.add(CargoRegistryCrateVersion.fromJson(line))
+                    } catch (e: Exception) {
+                        LOG.warn("${treeWalk.pathString}: ${e.message}")
+                    }
+                }
+
+                crates.put(treeWalk.nameString, CargoRegistryCrate(versions))
+            }
+        }
+        crates.force()
+    }
+
+    fun invalidateCaches() {
+        corruptionMarkerFile.apply {
+            parent?.createDirectories()
+            Files.createFile(this)
+        }
+    }
+
+    override fun dispose() {
+        crates.close()
+    }
+
+    private inner class CratesLocalIndexUpdateTask : Task.Backgroundable(null, "Updating cargo registry index", false) {
+        override fun run(indicator: ProgressIndicator) {
+            val branch = repository.resolve("origin/master") ?: return
+            val revCommit = RevWalk(repository).parseCommit(ObjectId.fromString(branch.name))
+            reloadCrates(revCommit.tree)
+        }
+
+        override fun onSuccess() {
+            state.indexedCommitHash = repository.resolve("origin/master")?.name ?: return
```

Comment:
There are two problems:
1. `indexedCommitHash` is not volatile, but there are write accesses from different threads
2. Last repository's commit might be fresher than `registryHeadCommitHash`, and so `state.indexedCommitHash != registryHeadCommitHash` might be always true (in `updateIfNeeded`)

Reply:
So I'd recommend to use
```suggestion
            state = CratesLocalIndexState(registryHeadCommitHash)
```